### PR TITLE
Address issues with saved report filters (1 of 2)

### PR DESCRIPTION
**Pull request by orangejenny**
##### SUMMARY
https://dimagi-dev.atlassian.net/browse/HI-878
https://dimagi-dev.atlassian.net/browse/HI-730
https://dimagi-dev.atlassian.net/browse/HI-814

Presumably introduced by the select2 migration in 8e086c1a752f84a1e87a83f7eac496273fbc54ab

One of the core change in the select2 upgrade was that multiselects now express their values as arrays of strings, whereas they used to be a single comma-separated string. This PR fixes a number of bugs related to code either treating a list as a string or vice versa, and it adds a migration to turn all of the current delimited-string filters into lists.

##### FEATURE FLAG
Some of this is specific to UCRs.

##### PRODUCT DESCRIPTION
This fixes several problems with multi-select report filters not working properly, particularly in saved reports, in both standard reports and UCRs.

### CODE REVIEW

File: corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+from corehq.apps.saved_reports.models import ReportConfig
+from corehq.apps.userreports.reports.filters.values import CHOICE_DELIMITER
+
+
+class Command(BaseCommand):
+    """
+        Migrated ReportConfig filters with multiple values from CHOICE_DELIMITER-delimited strings
+        to lists. This command can be deleted once that migration is complete.
+    """
```
Comment (czue) [inline corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py]: Is this the type of thing we should put in a django migration so other envs / open source users get the fix? Or we don't care about providing that level of support for this feature.

File: corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    def handle(self, *args, **options):
+        key = ["name slug"]
+        results = cache_core.cached_view(
```
Comment (czue) [inline corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py]: Curious why you're using the cached view and not going straight to couch?

File: corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            reduce=False,
+            include_docs=True,
+            startkey=key,
+            endkey=key + [{}]
+        )
+        count = 0
+        for result in results:
+            dirty = False
+            doc = result['doc']
+            config = ReportConfig.get(doc['_id'])
```
Comment (czue) [inline corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py]: This goes back to the DB to get the doc, no? Why can't you just call `.wrap` since you're already including docs? Or maybe better yet, use `iter_docs`?

File: corehq/apps/reports/filters/base.py [hunk 4]
```diff
@@ -169,7 +170,10 @@ class BaseMultipleOptionFilter(BaseSingleOptionFilter):
 
     @classmethod
     def get_value(cls, request, domain):
-        return request.GET.getlist(cls.slug)
+        selected_ids = []
+        for ids in request.GET.getlist(cls.slug):
+            selected_ids.extend(ids.split(CHOICE_DELIMITER))
```
Comment (czue) [inline corehq/apps/reports/filters/base.py]: why is this change necessary if we're now passing a list?

File: corehq/apps/reports/migrations/0003_multiselect_report_filters_are_lists.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        config = ReportConfig.get(doc['_id'])
+        for name, value in config['filters'].items():
+            if isinstance(value, str) and CHOICE_DELIMITER in value:
+                print("Updating config {} filter {}".format(config._id, name))
+                config['filters'][name] = value.split(CHOICE_DELIMITER)
+                dirty = True
+        if dirty:
+            count = count + 1
+            config.save()
+    print("Updated {} configs".format(count))
```
Comment (czue) [inline corehq/apps/reports/migrations/0003_multiselect_report_filters_are_lists.py]: (ignore the first question on the old file, but I think the couch questions are still valid)

File: corehq/apps/userreports/reports/view.py [hunk 6]
```diff
@@ -121,7 +121,11 @@ def query_dict_to_dict(query_dict, domain, string_type_params):
     for key in string_type_params:
         u_key = str(key)  # QueryDict's key/values are unicode strings
         if u_key in query_dict:
-            request_dict[key] = query_dict[u_key]  # json_request converts keys to strings
+            vals = query_dict.getlist(u_key)
+            if len(vals) > 1:
+                request_dict[key] = vals
+            else:
+                request_dict[key] = vals[0]
```
Comment (czue) [inline corehq/apps/userreports/reports/view.py]: Trying to think if there's any place this assumption wouldn't be valid. I took a more surgical approach in https://github.com/dimagi/commcare-hq/pull/25371/commits/3175af40eb7b49900e1d2698b8f846d26d856248

But I think this is probably ok.

**Review by czue — COMMENTED** [review]
Looks good - had some questions/suggestions on the migration that will be lost on the move but hopefully translate clearly.

One thing I'm unclear about is why the backend changes to support delimited strings are necessary.

File: corehq/apps/reports/filters/base.py (hunk 4 above)
Comment (czue) [reply corehq/apps/reports/filters/base.py]: this is answered in https://github.com/dimagi/commcare-hq/pull/25377

File: corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py]: dbb5fa96944

Comment (orangejenny) [reply corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py]: dbb5fa96944

File: corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py (hunk 3 above)
Comment (orangejenny) [reply corehq/apps/saved_reports/management/commands/fix_saved_reports_filters.py]: Switched to `wrap` in 6a8e7db400e

**Review by czue — APPROVED** [review]

File: corehq/apps/userreports/reports/view.py (hunk 6 above)
Comment (orangejenny) [reply corehq/apps/userreports/reports/view.py]: Frankly I just took the work that had already been done on this ticket and tested it, I'm fairly unfamiliar with UCRs and wasn't eager to make more changes in this code. I agree your PR is a more deliberate change, if you want to deploy it on top of this (given that you're deploying the other cleanup work in that branch) please go for it.

Comment (czue) [reply corehq/apps/userreports/reports/view.py]: Reading [the docs for `getlist`](https://docs.djangoproject.com/en/2.2/ref/request-response/#django.http.QueryDict.getlist) I think this should be safe and I like that it's much simpler than my version so seems good. I was just being cautious since I suspect these filters don't have great test coverage.

File: corehq/apps/reports/migrations/0003_multiselect_report_filters_are_lists.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+from django.db import migrations
+
+from corehq.apps.saved_reports.models import ReportConfig
+from corehq.apps.userreports.reports.filters.values import CHOICE_DELIMITER
+
+
+def _migrate_report_filters(apps, schema_editor):
+    """
+        Migrates ReportConfig filters with multiple values from CHOICE_DELIMITER-delimited strings to lists.
```
Comment (sravfeyn) [inline corehq/apps/reports/migrations/0003_multiselect_report_filters_are_lists.py]: 👍 Nice!

**Comment by sravfeyn** [comment]
Thanks for these, Jenny!
